docs/design/jet/025-jet-job-partition-pruning.md

```diff
@@ -0,0 +1,196 @@
+
+```
+- HASH JOIN
+    - SCAN[map1, partitionKey=1]
+    - SCAN[map2, partitionKey=K]
+```
+
+![Scan + Scan -> HashJoin](https://svgshare.com/i/tfk.svg)
+
+##### Member pruning for Scan -> Transform -> Aggregate
+
+```
+- AGGREGATION
+  - TRANSFORM
+    - SCAN[map, partitionKey=K]
+```
+
+![Scan + Scan -> HashJoin](https://svgshare.com/i/tTL.svg)
+
+On the picture above the basic units are processors, since the actual graph looks like `Scan -> Transform -> Aggregate`.
+So, here we can eliminate whole Member 2, since according to partition key condition only Member 1 contains required data
+and final destination point `Aggregate`. Moreover, it is a good example for further processor pruning optimization.
+
+#### Solution design details
+
+There are two different approaches were considered, which differ from each other
+in the stage that we understand that some member may be pruned from job execution.
+The _generic_ approach tries to define the default behavior and applicable to DAGs from any source (SQL/Jet),
+whereas the _SQL-oriented_ approach focuses on hinting `ExecutionPlanBuilder` with
+additional meta-information from `JobConfig` and DAG constructed by SQL optimizer.
+
+##### Generic approach
+
+The core idea here is that `ExecutionPlanBuilder` would perfrom a detailed analysis of the received DAG and possibly
+try to optimize it before creating execution plans. The requirements to prune a member are:
+
+- each vertex in the DAG may work without an input;
+- DAG does not contain `distributed-broadcast` edges;
+
+This approach generifies member pruning for SQL, Pipeline API and any connector.
+However, it brings high complexity in problem analysis, long development time
+and big list of corner cases, which are discussed in the chapter below.
+
+##### SQL-oriented approach
+
+Unlike the way described above, we can think that member pruning would be beneficial
+mostly for small jobs. The overwhelming majority of such jobs are generated by SQL,
+and our team effort is to replace Predicate API, where `PartitionPredicate` is available
+for partitioned data querying. For that, we will use SQL optimization phase also to determine
+partitions and for all relations. Then, we extract required partitions set to run to `JobConfig`
+baked by `__sql.requiredPartitions` argument. This indicates that `ExecutionPlanBuilder` will
+try to choose only required members.
+
+### Scan partition pruning
+
+Partition pruning for IMap scan is a pretty simple optimization: we will extract partition key condition during SQL opt phase,
+pass it to specialized processor supplier which will spawn `ReadMapOrCacheP` with only required partitions to scan.
+
+### `lazyForceTotalParallelismOne`
+
+To complement the existing `forceTotalParallelismOne` with PMS with partition pruning support we decided
+to add new partition-aware `lazyForceTotalParallelismOne` PMS builder which does not cache member address
```

Comment:
it not only "does not cache" but also calculates member dynamically based on partition key _expressions_ - this "detail" is completely missing from TDD. You we also supposed to describe shortly how can that be used by SQL.

Reply:
from previous comment:
>lazyForceTotalParallelismOne is not only lazy, but the main characteristic is that is it partition-assignment aware. you completelty skipped description of that while this is quite important